README.md
# djangocms_ads
Google adverts for django-cms


HISTORY.rst
=======
History
=======

`Unreleased`_
=============

`1.1.2`_ (2023-08-31)
=====================

* Fix migration dependency from django-cms due to new migrations in django-cms v3. `#32`_

`1.1.1`_ (2023-03-24)
=====================

* Fix migration dependency from django-cms. `#21`_

`1.1.0`_ (2022-07-27)
=====================

* Added python 3.10 support. `#11`_

`1.0.1`_ (2022-03-02)
=====================

* Added ``py.typed`` and removed v4 cms restriction `#1`_

`1.0.0`_ (2022-03-02)
=====================

* First release.

.. _1.0.0: https://github.com/marksweb/djangocms-ads/tree/1.0.0
.. _1.0.1: https://github.com/marksweb/djangocms-ads/compare/1.0.0...1.0.1
.. _1.1.0: https://github.com/marksweb/djangocms-ads/compare/1.0.1...1.1.0
.. _1.1.1: https://github.com/marksweb/djangocms-ads/compare/1.1.0...1.1.1
.. _Unreleased: https://github.com/marksweb/djangocms-ads/compare/1.1.1...master

.. _#1: https://github.com/marksweb/djangocms-ads/pull/1
.. _#11: https://github.com/marksweb/djangocms-ads/pull/11
.. _#21: https://github.com/marksweb/djangocms-ads/pull/21
.. _#32: https://github.com/marksweb/djangocms-ads/pull/32


test_settings.py
from __future__ import annotations

HELPER_SETTINGS = {
    "TEST_RUNNER": "app_helper.pytest_runner.PytestTestRunner",
    "SECRET_KEY": "djangocms-ads-test-suite",
    "USE_TZ": False,
    "TIME_ZONE": "Europe/London",
    "INSTALLED_APPS": [
        "djangocms_attributes_field",
        "djangocms_ads",
    ],
    "MIGRATION_MODULES": {
        "auth": None,
        "cms": None,
        "menus": None,
        "djangocms_ads": None,
    },
    "CMS_PERMISSION": True,
    "LANGUAGES": (
        ("en", "English"),
    ),
    "CMS_LANGUAGES": {
        1: [
            {"code": "en", "name": "English"},
        ],
    },
    "LANGUAGE_CODE": "en",
    "DEFAULT_AUTO_FIELD": "django.db.models.AutoField",
}


def run() -> None:
    from app_helper import runner

    runner.cms("djangocms_ads", extra_args=[])


if __name__ == "__main__":
    run()


tests/test_models.py
from __future__ import annotations

from django.test import TestCase
from djangocms_ads.models import (
    AdSlotPlugin,
    AdvancedAdContainerPlugin,
    AdvertPlugin,
    SimpleAdContainerPlugin,
)

SCRIPT = \
"""
<script>
  window.googletag = window.googletag || {cmd: []};
  googletag.cmd.push(function() {
      googletag.defineSlot('/4321/Leaderboard', [728, 90], 'div-gpt-ad-1232449682314-0').addService(googletag.pubads());
      googletag.defineSlot('/4321/MPU', [300, 250], 'div-gpt-ad-1523479823898-0').addService(googletag.pubads());
    googletag.pubads().enableSingleRequest();
    googletag.pubads().collapseEmptyDivs();
    googletag.enableServices();
  });
</script>
""" # noqa


class ModelTestCase(TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_advertplugin_instance(self):
        AdvertPlugin.objects.create(
            label="test advert",
            div_id="test-id-123",
        )
        instance = AdvertPlugin.objects.all()
        self.assertEqual(instance.count(), 1)
        instance = AdvertPlugin.objects.first()
        self.assertEqual(instance.label, "test advert")
        self.assertEqual(instance.div_id, "test-id-123")
        # test string repr
        self.assertEqual(str(instance), "test advert")

    def test_advanced_instance(self):
        AdvancedAdContainerPlugin.objects.create(
            internal_name="Test advanced container",
            content=SCRIPT,
        )
        instance = AdvancedAdContainerPlugin.objects.all()
        self.assertEqual(instance.count(), 1)
        instance = AdvancedAdContainerPlugin.objects.first()
        # test strings
        self.assertEqual(str(instance), "Test advanced container")
        self.assertEqual(instance.content, SCRIPT)

    def test_simple_instance(self):
        SimpleAdContainerPlugin.objects.create(
            internal_name="Test simple container",
        )
        instance = SimpleAdContainerPlugin.objects.all()
        self.assertEqual(instance.count(), 1)
        instance = SimpleAdContainerPlugin.objects.first()
        # test strings
        self.assertEqual(str(instance), "Test simple container")

    def test_adslot_instance(self):
        AdSlotPlugin.objects.create(
            internal_name="Test ad slot",
            ad_unit_path="/1234/testing",
            sizes="[10, 20]",
            div_id="test-ad-1234"
        )
        instance = AdSlotPlugin.objects.all()
        self.assertEqual(instance.count(), 1)
        instance = AdSlotPlugin.objects.first()
        # test strings
        self.assertEqual(str(instance), "Test ad slot")


tests/test_plugins.py
from __future__ import annotations

from cms.api import add_plugin, create_page
from cms.test_utils.testcases import CMSTestCase

SCRIPT = \
"""
<script>
  window.googletag = window.googletag || {cmd: []};
  googletag.cmd.push(function() {
      googletag.defineSlot('/4321/Leaderboard', [728, 90], 'div-gpt-ad-1232449682314-0').addService(googletag.pubads());
      googletag.defineSlot('/4321/MPU', [300, 250], 'div-gpt-ad-1523479823898-0').addService(googletag.pubads());
    googletag.pubads().enableSingleRequest();
    googletag.pubads().collapseEmptyDivs();
    googletag.enableServices();
  });
</script>
"""  # noqa


class AdPluginsTestCase(CMSTestCase):

    def setUp(self):
        self.language = "en"
        self.page = create_page(
            title="home",
            template="page.html",
            language=self.language,
        )
        # self.page.publish(self.language)
        self.placeholder = self.page.placeholders.get(slot="content")
        self.superuser = self.get_superuser()

    def tearDown(self):
        self.page.delete()
        self.superuser.delete()

    def test_advanced_ad_rendering(self):
        request_url = self.page.get_absolute_url(
            self.language) + "?toolbar_off=true"

        add_plugin(
            self.page.placeholders.get(slot="content"),
            "AdvancedAdContainer",
            self.language,
            internal_name="advanced plugin",
            content=SCRIPT,
            template="default",
        )
        advert = add_plugin(
            self.page.placeholders.get(slot="content"),
            "Advert",
            self.language,
            label="Test render advert",
            template="default",
            div_id="div-gpt-ad-1232449682314-0"
        )

        self.page.publish(self.language)
        self.assertEqual(advert.label, "Test render advert")
        self.assertEqual(advert.div_id, "div-gpt-ad-1232449682314-0")

        with self.login_user_context(self.superuser):
            response = self.client.get(request_url)

        self.assertIn(b"div-gpt-ad-1232449682314-0", response.content)

    def test_simple_ad_rendering(self):
        request_url = self.page.get_absolute_url(
            self.language) + "?toolbar_off=true"

        add_plugin(
            self.page.placeholders.get(slot="content"),
            "SimpleAdContainer",
            self.language,
            internal_name="simple plugin",
            template="default",
        )
        ad_slot = add_plugin(
            self.page.placeholders.get(slot="content"),
            "AdSlot",
            self.language,
            internal_name="render test slot",
            template="default",
            ad_unit_path="/4321/Leaderboard",
            sizes="[300, 250]",
            div_id="div-render-test"
        )
        advert = add_plugin(
            self.page.placeholders.get(slot="content"),
            "Advert",
            self.language,
            label="Test render advert",
            template="default",
            div_id="div-render-test"
        )

        self.page.publish(self.language)
        self.assertEqual(ad_slot.div_id, "div-render-test")
        self.assertEqual(advert.label, "Test render advert")
        self.assertEqual(advert.div_id, "div-render-test")

        with self.login_user_context(self.superuser):
            response = self.client.get(request_url)

        self.assertIn(b"div-render-test", response.content)


requirements/compile.py
#!/usr/bin/env python
from __future__ import annotations

import os
import subprocess
import sys
from pathlib import Path

if __name__ == "__main__":
    os.chdir(Path(__file__).parent)
    os.environ["CUSTOM_COMPILE_COMMAND"] = "requirements/compile.py"
    os.environ.pop("PIP_REQUIRE_VIRTUALENV", None)
    common_args = [
        "-m",
        "piptools",
        "compile",
        "--generate-hashes",
        "--allow-unsafe",
    ] + sys.argv[1:]
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=2.2,<2.3",
            "--upgrade-package",
            "django-cms==3.7",
            "-o",
            "py38-cms37-django22.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=2.2,<2.3",
            "--upgrade-package",
            "django-cms==3.7",
            "-o",
            "py38-cms37-django31.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=2.2,<2.3",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py38-cms39-django22.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=3.1,<3.2",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py38-cms39-django31.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=3.2,<4.0",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py38-cms39-django32.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.9",
            *common_args,
            "--upgrade-package",
            "Django>=2.2,<2.3",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py39-cms39-django22.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.9",
            *common_args,
            "--upgrade-package",
            "Django>=3.1,<3.2",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py39-cms39-django31.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.9",
            *common_args,
            "--upgrade-package",
            "Django>=3.2a1,<3.3",
            "--upgrade-package",
            "django-cms==3.9",
            "-o",
            "py39-cms39-django32.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.8",
            *common_args,
            "--upgrade-package",
            "Django>=3.2,<4.0",
            "--upgrade-package",
            "django-cms==3.10",
            "-o",
            "py38-cms310-django32.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.9",
            *common_args,
            "--upgrade-package",
            "Django>=3.2,<4.0",
            "--upgrade-package",
            "django-cms==3.10",
            "-o",
            "py39-cms310-django32.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.10",
            *common_args,
            "--upgrade-package",
            "Django>=3.2,<4.0",
            "--upgrade-package",
            "django-cms==3.10",
            "-o",
            "py310-cms310-django32.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.11",
            *common_args,
            "--upgrade-package",
            "Django>=4.1,<4.2",
            "--upgrade-package",
            "django-cms==3.11",
            "-o",
            "py311-cms311-django41.txt",
        ],
        check=True,
        capture_output=True,
    )
    subprocess.run(
        [
            "python3.11",
            *common_args,
            "--upgrade-package",
            "Django>=4.2rc1,<5.0",
            "--upgrade-package",
            "django-cms==3.11",
            "-o",
            "py311-cms311-django42.txt",
        ],
        check=True,
        capture_output=True,
    )


src/djangocms_ads/models.py
from __future__ import annotations

from typing import Any

from cms.models import CMSPlugin
from django.conf import settings
from django.db import models
from django.utils.translation import gettext_lazy as _
from djangocms_attributes_field.fields import AttributesField

__all__ = [
    "AdSlotPlugin",
    "AdvancedAdContainerPlugin",
    "AdvertPlugin",
    "SimpleAdContainerPlugin"
]


# Add additional choices through the ``settings.py``.
TEMPLATE_DEFAULT = "default"


def get_templates() -> list[tuple[str, Any]]:
    choices = [(TEMPLATE_DEFAULT, _("Default"))]
    choices += getattr(settings, "DJANGOCMS_ADS_TEMPLATES", [])
    return choices


class AdvancedAdContainerPlugin(CMSPlugin):
    """Advanced ad container allowing javascript to be added directly"""
    internal_name = models.CharField(
        verbose_name=_("internal name"),
        max_length=255,
        help_text=_("Provide internal name for Ad slot."),
    )
    content = models.TextField(
        verbose_name=_("Advert script tag"),
        help_text=_("The script tag for the head of the page")
    )
    template = models.CharField(
        verbose_name=_("Template"),
        choices=get_templates(),
        default=TEMPLATE_DEFAULT,
        max_length=255,
    )
    modified = models.DateTimeField(
        verbose_name=_("Modified"), auto_now=True
    )

    class Meta:
        app_label = "djangocms_ads"
        verbose_name = _("Advanced ad container")
        verbose_name_plural = _("Advanced ad containers")

    def __str__(self) -> str:
        return self.internal_name


class SimpleAdContainerPlugin(CMSPlugin):
    """Simple ad container which doesn't contain javascript"""
    internal_name = models.CharField(
        verbose_name=_("internal name"),
        max_length=255,
        help_text=_("Provide internal name for Ad slot."),
    )
    template = models.CharField(
        verbose_name=_("Template"),
        choices=get_templates(),
        default=TEMPLATE_DEFAULT,
        max_length=255,
    )
    modified = models.DateTimeField(
        verbose_name=_("Modified"), auto_now=True
    )

    class Meta:
        app_label = "djangocms_ads"
        verbose_name = _("Simple ad container")
        verbose_name_plural = _("Simple ad containers")

    def __str__(self) -> str:
        return self.internal_name


class AdSlotPlugin(CMSPlugin):
    """Ad slot definition"""
    internal_name = models.CharField(
        verbose_name=_("internal name"),
        max_length=255,
        help_text=_("Provide internal name for Ad slot."),
    )
    ad_unit_path = models.CharField(
        verbose_name=_("Ad unit path"),
        max_length=255,
        help_text=_(
            "The first item provided by google "
            "for the slot, e.g. '/4321234/MPU'"
        )
    )
    sizes = models.CharField(
        verbose_name=_("Ad sizes"),
        max_length=255,
        help_text=_(
            "The second item provided by google for the "
            "slot, e.g. [300, 250]"
        )
    )
    div_id = models.CharField(
        verbose_name=_("Div ID"),
        max_length=100,
        help_text=_(
            "The third item provided by google for the slot, "
            "e.g. 'div-gpt-ad-12345-0'"
        )
    )
    template = models.CharField(
        verbose_name=_("Template"),
        choices=get_templates(),
        default=TEMPLATE_DEFAULT,
        max_length=255,
    )
    modified = models.DateTimeField(
        verbose_name=_("Modified"), auto_now=True
    )

    class Meta:
        app_label = "djangocms_ads"
        verbose_name = _("ad slot")
        verbose_name_plural = _("ad slots")

    def __str__(self) -> str:
        return self.internal_name


class AdvertPlugin(CMSPlugin):
    """Advert to display in the page"""
    label = models.CharField(verbose_name=_("label"), max_length=120)
    template = models.CharField(
        verbose_name=_("Template"),
        choices=get_templates(),
        default=TEMPLATE_DEFAULT,
        max_length=255,
    )
    div_id = models.CharField(
        verbose_name=_("Div ID"),
        max_length=100,
        help_text=_(
            "The third item provided by google for the "
            "slot, e.g. 'div-gpt-ad-12345-0'"
        )
    )
    attributes = AttributesField(
        verbose_name=_("Attributes"),
        blank=True,
        excluded_keys=["href", "target", "id"]
    )

    class Meta:
        app_label = "djangocms_ads"
        verbose_name = _("advert plugin model")
        verbose_name_plural = _("advert plugin models")

    def __str__(self) -> str:
        return self.label


src/djangocms_ads/cms_plugins.py
from __future__ import annotations

from typing import Any

from cms.models import Placeholder
from cms.plugin_base import CMSPluginBase
from cms.plugin_pool import plugin_pool
from django.utils.translation import gettext_lazy as _

from .models import (
    AdSlotPlugin,
    AdvancedAdContainerPlugin,
    AdvertPlugin,
    SimpleAdContainerPlugin,
)

__all__ = ["Advert"]


@plugin_pool.register_plugin
class AdvancedAdContainer(CMSPluginBase):
    """Advanced ad container plugin"""
    module = _("Adverts")
    name = _("Advert Container (advanced)")
    model = AdvancedAdContainerPlugin

    def get_render_template(
        self, context: dict[Any, Any], instance: AdvancedAdContainerPlugin,
            placeholder: Placeholder) -> str:
        return f"djangocms_ads/{instance.template}/container_advanced.html"


@plugin_pool.register_plugin
class SimpleAdContainer(CMSPluginBase):
    """Simple ad container plugin"""
    module = _("Adverts")
    name = _("Advert Container")
    model = SimpleAdContainerPlugin
    allow_children = True
    child_classes = ["AdSlot"]

    def get_render_template(
        self, context: dict[Any, Any], instance: SimpleAdContainerPlugin,
            placeholder: Placeholder) -> str:
        return "djangocms_ads/default/container_simple.html"


@plugin_pool.register_plugin
class AdSlot(CMSPluginBase):
    """Ad slot plugin"""
    module = _("Adverts")
    name = _("Advert Slot")
    model = AdSlotPlugin
    require_parent = True
    parent_classes = ["SimpleAdSlotContainer"]

    def get_render_template(
        self, context: dict[Any, Any], instance: AdSlotPlugin,
            placeholder: Placeholder) -> str:
        return f"djangocms_ads/{instance.template}/ad_slot.html"


@plugin_pool.register_plugin
class Advert(CMSPluginBase):
    """Advert plugin"""
    module = _("Adverts")
    name = _("Advert")
    model = AdvertPlugin

    def get_render_template(
        self, context: dict[Any, Any], instance: AdvertPlugin,
            placeholder: Placeholder) -> str:
        return f"djangocms_ads/{instance.template}/advert.html"


src/djangocms_ads/migrations/0001_initial.py
# Generated by Django 3.2.12 on 2022-02-28 23:34

from __future__ import annotations

import django.db.models.deletion
from django.db import migrations, models

import djangocms_attributes_field.fields


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('cms', '__first__'),
    ]

    operations = [
        migrations.CreateModel(
            name='AdSlotPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='djangocms_ads_adslotplugin', serialize=False, to='cms.cmsplugin')),
                ('internal_name', models.CharField(help_text='Provide internal name for Ad slot.', max_length=255, verbose_name='internal name')),
                ('ad_unit_path', models.CharField(help_text="The first item provided by google for the slot, e.g. '/4321234/MPU'", max_length=255, verbose_name='Ad unit path')),
                ('sizes', models.CharField(help_text='The second item provided by google for the slot, e.g. [300, 250]', max_length=255, verbose_name='Ad sizes')),
                ('div_id', models.CharField(help_text="The third item provided by google for the slot, e.g. 'div-gpt-ad-12345-0'", max_length=100, verbose_name='Div ID')),
                ('template', models.CharField(choices=[('default', 'Default')], default='default', max_length=255, verbose_name='Template')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='Modified')),
            ],
            options={
                'verbose_name': 'ad slot',
                'verbose_name_plural': 'ad slots',
            },
            bases=('cms.cmsplugin',),
        ),
        migrations.CreateModel(
            name='AdvancedAdContainerPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='djangocms_ads_advancedadcontainerplugin', serialize=False, to='cms.cmsplugin')),
                ('internal_name', models.CharField(help_text='Provide internal name for Ad slot.', max_length=255, verbose_name='internal name')),
                ('content', models.TextField(help_text='The script tag for the head of the page', verbose_name='Advert script tag')),
                ('template', models.CharField(choices=[('default', 'Default')], default='default', max_length=255, verbose_name='Template')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='Modified')),
            ],
            options={
                'verbose_name': 'Advanced ad container',
                'verbose_name_plural': 'Advanced ad containers',
            },
            bases=('cms.cmsplugin',),
        ),
        migrations.CreateModel(
            name='AdvertPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='djangocms_ads_advertplugin', serialize=False, to='cms.cmsplugin')),
                ('label', models.CharField(max_length=120, verbose_name='label')),
                ('template', models.CharField(choices=[('default', 'Default')], default='default', max_length=255, verbose_name='Template')),
                ('div_id', models.CharField(help_text="The third item provided by google for the slot, e.g. 'div-gpt-ad-12345-0'", max_length=100, verbose_name='Div ID')),
                ('attributes', djangocms_attributes_field.fields.AttributesField(blank=True, default=dict, verbose_name='Attributes')),
            ],
            options={
                'verbose_name': 'advert plugin model',
                'verbose_name_plural': 'advert plugin models',
            },
            bases=('cms.cmsplugin',),
        ),
        migrations.CreateModel(
            name='SimpleAdContainerPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='djangocms_ads_simpleadcontainerplugin', serialize=False, to='cms.cmsplugin')),
                ('internal_name', models.CharField(help_text='Provide internal name for Ad slot.', max_length=255, verbose_name='internal name')),
                ('template', models.CharField(choices=[('default', 'Default')], default='default', max_length=255, verbose_name='Template')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='Modified')),
            ],
            options={
                'verbose_name': 'Simple ad container',
                'verbose_name_plural': 'Simple ad containers',
            },
            bases=('cms.cmsplugin',),
        ),
    ]
